### DISCUSSION TITLE
Question about phase field fracture model for multi-phase material.

### DISCUSSION THREAD

**Discussion by Yaode-Yin** [idaholab/moose#27921] [Q&A Modules: Phase field]
Hello Moose expert:
  I try to use Moose to simulate the fracture of composite materials, and I find it convenient to introduce an additional order-parameter eta to distinguish the matrix and reinforcement of composites (eta = 0 for matrix, and eta = 1 for reinforcement). Let’s start form the simplest example: one circle reinforcement material under transverse loading at plane-strain state, as the figure.
 
![image](https://github.com/idaholab/moose/assets/146060023/66656538-0de6-4b0c-813d-aa4a2133eb3e)

  I define `base_name = A` or `base_name = B `to distinguish the two materials in Moose material system, and then use `SwitchingFunctionalMaterial` with the order parameter eta to finish the composite material model. But errors occur when the input file is submitted.

```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

The following material properties are declared on block 0 by multiple materials:
  Material Property             Material Objects
  d2Fdcdstrain                  combined damage_stress_A damage_stress_B degradation fracture_driving_energy local_fracture_energy mech_strain switching 
  hist                          combined damage_stress_A damage_stress_B degradation fracture_driving_energy local_fracture_energy mech_strain switching
```
Here are the codes for the Moose material system:
```
[Materials]
  [./pfbulkmat]
    type = GenericConstantMaterial
    prop_names = 'gc_prop l visco'
    prop_values = '2.7 0.1 1e-32'
  [../]
  [./define_mobility]
    type = ParsedMaterial
    material_property_names = 'gc_prop visco'
    property_name = L
    expression = '1.0/(gc_prop * visco)'
  [../]
  [./define_kappa]
    type = ParsedMaterial
    material_property_names = 'gc_prop l'
    property_name = kappa_op
    expression = '3 * gc_prop * l /4'
  [../]
  [./elastic_tensor_A]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 100e3
    poissons_ratio = 0.3
     base_name = A
  [../]
  [./strain_A]
    type = ComputeSmallStrain
    base_name = A
  [../]
  [./elastic_tensor_B]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 10e3
    poissons_ratio = 0.3
     base_name = B
  [../]
    [./strain_B]
    type = ComputeSmallStrain
    base_name = B
  [../]
  [./damage_stress_A]
    type = ComputeLinearElasticPFFractureStress
    c = c
    E_name = 'elastic_energy_A'
    D_name = 'degradation'
    F_name = 'local_fracture_energy'
    decomposition_type = none
    use_snes_vi_solver = true
    base_name = A
  [../]
   [./damage_stress_B]
    type = ComputeLinearElasticPFFractureStress
    c = c
    E_name = 'elastic_energy_B'
    D_name = 'degradation'
    F_name = 'local_fracture_energy'
    decomposition_type = none
    use_snes_vi_solver = true
    base_name = B
  [../]
##########################################3
  [./degradation]
    type = DerivativeParsedMaterial
    property_name = degradation
    coupled_variables = 'c'
    expression = '(1-c)^2*(1-eta)+eta'
    constant_names       = 'eta'
    constant_expressions = '0.0'
    derivative_order = 2
  [../]
  [./local_fracture_energy]
    type = DerivativeParsedMaterial
    property_name = local_fracture_energy
    coupled_variables = 'c'
    material_property_names = 'gc_prop l'
    expression = '3* c * gc_prop / 8 / l'
    derivative_order = 2
  [../]
  [./fracture_driving_energy]
    type = DerivativeSumMaterial
    coupled_variables = c
    sum_materials = 'elastic_energy_A elastic_energy_B  local_fracture_energy'
    derivative_order = 2
    property_name = F
  [../]
##############################################
  [./switching]
    type = SwitchingFunctionMaterial
    eta = eta
  [../]
 [./combined]
    type = TwoPhaseStressMaterial
    base_A = A
    base_B =B
  [../]
[]

```
Using the `Debug` system to set `show_material_props = true`, I find `ComputeLinearElasticPFFractureStress` calculate` d2Fdcdstrain `and `hist` without the mentioned base name A or B. That’s why the calculation fails. How should I modify to correctly model the phase field fracture of a composite material?

Best wishes!

**Comment by GiudGiud** [comment]
Hello

It looks like a bug to me.
Feel free to modify `ComputeLinearElasticPFFractureStress` and add `_base_name` in the declare class, like this:
```
declareProperty<Real>(_base_name + "hist")
```

We ll wait until someone from the phase field community pitches in to be sure

Guillaume

**Reply by Yaode-Yin** [reply]
Thanks for your reply. I looked at the source code of [ComputeLinearElasticPFFractureStress](https://mooseframework.inl.gov/docs/doxygen/modules/classComputeLinearElasticPFFractureStress.html#details) and find that `d2Fdcdstrain` is probably wrong. It should be `d2Edcdstrain` , where E stands for the elastic energy.
![image](https://github.com/idaholab/moose/assets/146060023/ecd9c492-81a5-4c4e-bdac-02b116f18181)

I'm still checking on the `hist` issue.

**Reply by Yaode-Yin** [reply]
I find a simpe way to fix the problems, but dont ensure it suitable for all cases. Just replace `d2Fdcdstrain` by `d2Edcdstrain` in servral .c or .h files included  `d2Fdcdstrain`.  As for `hist` issue,   just modify `ComputeLinearElasticPFFractureStress` and add ` _base_name`  in the declare class. 
Because I use the variational inequality solver, I dont need to face the issues about  `hist` or `d2Fdcdstrain` directly. So I think the simple modification is effective.

**Reply by GiudGiud** [reply]
why do you not think the `d2F/Edcdstrain` should be prefaced with _base_name?

**Reply by Yaode-Yin** [reply]
Yes, `d2F/Edcdstrain` also needs to be prefaced with _base_name. Otherwise, the claculation cannot strat.
